pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,365 @@
+	for key, o := range t.observed {
+		if key == "" {
+			continue
+		}
+		keys := strings.Split(key, string(sep))
+
+		keys = append(keys, t.userID)
+		if o.activeSerie.Load() > 0 {
+			out <- prometheus.MustNewConstMetric(t.activeSeriesPerUserAttribution, prometheus.GaugeValue, o.activeSerie.Load(), keys...)
+		}
+		if o.receivedSample.Load() > 0 {
+			out <- prometheus.MustNewConstMetric(t.receivedSamplesAttribution, prometheus.CounterValue, o.receivedSample.Load(), keys...)
+		}
+		o.discardedSampleMtx.Lock()
+		for reason, discarded := range o.discardedSample {
+			out <- prometheus.MustNewConstMetric(t.discardedSampleAttribution, prometheus.CounterValue, discarded.Load(), append(keys, reason)...)
+		}
+		o.discardedSampleMtx.Unlock()
+	}
+}
+
+func (t *Tracker) IncrementDiscardedSamples(lbs []mimirpb.LabelAdapter, value float64, reason string, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCountersWithLabelAdapter(lbs, now, 0, 0, value, &reason, true)
+}
+
+func (t *Tracker) IncrementReceivedSamples(req *mimirpb.WriteRequest, now time.Time) {
+	if t == nil {
+		return
+	}
+
+	dict := make(map[string]int)
+	for _, ts := range req.Timeseries {
+		lvs := t.extractLabelValuesFromLabelAdapater(ts.Labels)
+		dict[t.hashLabelValues(lvs)] += len(ts.TimeSeries.Samples) + len(ts.TimeSeries.Histograms)
+	}
+	for k, v := range dict {
+		t.updateCountersCommon(k, now, 0, float64(v), 0, nil, true)
+	}
+}
+
+func (t *Tracker) updateCountersWithLabelAdapter(lbls []mimirpb.LabelAdapter, ts time.Time, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	labelValues := t.extractLabelValuesFromLabelAdapater(lbls)
+	key := t.hashLabelValues(labelValues)
+	t.updateCountersCommon(key, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+}
+
+func (t *Tracker) hashLabelValues(labelValues []string) string {
+	buf := bufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufferPool.Put(buf)
+
+	for i, value := range labelValues {
+		if i > 0 {
+			buf.WriteRune(sep)
+		}
+		buf.WriteString(value)
+	}
+	return buf.String()
+}
+
+func (t *Tracker) extractLabelValuesFromLabelAdapater(lbls []mimirpb.LabelAdapter) []string {
+	labelValues := make([]string, len(t.labels))
+	for idx, cal := range t.labels {
+		for _, l := range lbls {
+			if l.Name == cal {
+				labelValues[idx] = l.Value
+				break
+			}
+		}
+		if labelValues[idx] == "" {
+			labelValues[idx] = missingValue
+		}
+	}
+	return labelValues
+}
+
+func (t *Tracker) updateCounters(lbls labels.Labels, ts time.Time, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	labelValues := make([]string, len(t.labels))
+	for idx, cal := range t.labels {
+		labelValues[idx] = lbls.Get(cal)
+		if labelValues[idx] == "" {
+			labelValues[idx] = missingValue
+		}
+	}
+	key := t.hashLabelValues(labelValues)
```

Comment:
I am not so sure that get a key string everytime is a very good idea, but it did make the code less complex